### ISSUE TITLE
IPLD data structure shape consistency

### ISSUE THREAD

**Issue by mikeal** [ipld/ipld#66]
## Background

As we’ve been spending more time with IPLD some “best practices” are emerging. One of them has been designing for what I would call “shape consistency.” By this, I mean that the shape of the data structure is consistent even if the block boundaries are **not** consistent.

One problem we had in `unixfs-v2` was trying to design for several different and often competing use cases that would require us to vary how the blocks were constructed and linked together. Trying to do this at the spec level felt endless until we realized that we didn’t really **need** to do it this way. Instead, if we design the spec and the implementations to use path lookups for properties the block boundary differences are transparent. This allows applications to construct the blocks in whatever way is most performant for their use case.

As we’ve been going down the Selectors rabbit hole this practice is even more important. It appears that in our “IPLD Everywhere” future selectors will likely be the primary way we perform read operations on IPLD graphs. Designing things to be agnostic of the block boundary but with a consistent shape gives use the same benefits I outlined above in the `unixfs-v2` example.

## Advanced Data Structures (Collections, Indexes)

One question I’d like to raise is “Is varying the block boundary **always** enough flexibility to create the data structures we want? Could we actually impose a requirement on shape consistency?”

If the answer is “yes” then it makes a big difference in how we design dynamic resolution in more advanced data structures.

For instance, here’s two different interface paths we might take depending on whether or not the statement above is true.

### Inconsistent Shape

When traversal hits the root node of a collection we would lookup the resolution function for that collection and:

* Pass it the decoded root block and the query (either Key, or Range(start, end))
* It would return either
  * The value
  * The next CID to lookup along with any continuation data it needs passed into the next call.
* Each CID would be resolved and passed back into the function, along with the continuation data, until the value is reached.

### Consistent Shape

When traversal hits the root node of a collection we would lookup the resolution function for that collection and:

* Pass it the decoded root block and the query (either Key or Range(start, end))
* It would return a selector, probably with recursive attribute matching, that would would resolve to the value once traversed (or would end up not matching at some point, in which case the value does not actually exist in the collection).

As we start to implement string sorted collections we’re going to end up with btree-like structures, and the shape will vary quite a lot depend on the data in the tree. I’d like to figure out if, with an advanced enough selector engine, we could maintain consistent enough shapes that this would be enough.

It would make creating new data structures a lot easier, and much faster to resolve, but I’m a little worried that there are data structures we have no yet considered that would effectively be impossible to implement without a more dynamic resolution interface.

**Comment by rvagg** [comment]
@mikeal this is going to need a little more of the modelling that's in your head for it to make sense (to me at least).

Some immediate questions, although I'm aware that I'm not getting close enough to the core issues here yet.

1. A definition of "block boundaries" would be helpful, is this simply pointing to data spanning arbitrarily within and/or across actual blocks? So you might have a large structure that could go in a single block, or it might get split up across a bunch of blocks for some (perhaps arbitrary) reason?
2. Why would it matter if you traverse by CID or selector? Is it simply that a selector is a _nicer_ abstraction above CID-traversal and works both within and across blocks? The selector is still causing CID-traversal though, right? Perhaps in more compressed (in the sense of spanning fewer blocks) data structures it means traversing fewer blocks, but it's still traversing. So this only matters to the resolution function in that it doesn't need to understand about the concept of blocks, that's just pushed down a layer or two? 
3. In the "btree-like structures", you're imagining that given a root block, it could return a selector that points directionally down through its nodes to a possible end-value? "Yeah, that thing you want, it's roughly in that direction from here and here's some instructions for how you'd choose which leaf when you get to each node". Is this kind of like trying to put the intelligence into the selector rather than the resolution function? So the resolution function essentially only knows how to get a selector from a root node and passes that off to the selector-handling mechanism?

**Comment by mikeal** [comment]
Here’s a quick graphic I did as part of a talk. It shows how you can vary the way you create the blocks for a simple btree.

![4FE296A9-0117-41C7-9E9E-6B6E821FDB9D](https://user-images.githubusercontent.com/579/54404483-ce300800-4690-11e9-85ec-c046056cf546.png)

I’ll write up more detailed answers to the rest of your questions soon, I’m just a little too tired atm :)

**Comment by mikeal** [comment]
> Why would it matter if you traverse by CID or selector? Is it simply that a selector is a nicer abstraction above CID-traversal and works both within and across blocks? The selector is still causing CID-traversal though, right?

Correct, you don’t save any traversal effort. The main difference would be, if you can get a selector for the traversal that’s a single function call and the traversal engine has everything it needs. If you have to call a resolution function for each block using some kind of continuation passing you have a function execution per block during traversal. For many use cases, the performance difference is rather minimal, but when we are able to implement the traversal function in WebAssembly and attach it directly to the data structure calling it on every block may end up being expensive.

**Comment by Stebalien** [comment]
At the end of the day, applications _should_ be able to impose shape requirements on their data. That is, we need to expose some interface that acts like the current IPLD interfaces: one that handles blocks/fragments. This is a hard requirement to work with existing merkledag systems (e.g., git).

However, I'm not sure if I really understand your concern here. There's nothing we can do manually/dynamically that a sufficiently advanced selector engine can't do.

Of course, the key part there is sufficiently advanced. We _already_ have an issue with unixfs as it exists because we can't decode the unixfs protobuf from within a selector. However, in cases like this, we can still create a selector that selects everything we might need to perform the real query locally.

**Comment by mikeal** [comment]
I should have been clearer about the scope of my comment. I’m specifically thinking about **new** data structures we are building on the data model, not existing merkledag systems people are linking into.

**Comment by Stebalien** [comment]
I'm still confused what the concern is. We'll likely need special selectors (unfortunately) for traversing things like hamts, etc.
